File: 1900/1991.cpp
#include <bits/stdc++.h>
using namespace std;

class Tree{
private:
    int N;
    vector<int> lc;
    vector<int> rc;
public:
    Tree(): N(0){}
    Tree(int n): N(n){
        lc.resize(n,-1);
        rc.resize(n,-1);

    }
    void add_sub_tree(int p, int l, int r){
        if(l>0) lc[p] = l;
        if(r>0) rc[p] = r;
    }
    void preorder(int x){
        cout << (char)(x+65);
        if(lc[x] != -1) preorder(lc[x]);
        if(rc[x] != -1) preorder(rc[x]);
    }
    void inorder(int x){
        if(lc[x] != -1) inorder(lc[x]);
        cout << (char)(x+65);
        if(rc[x] != -1) inorder(rc[x]);
    }
    void postorder(int x){
        if(lc[x] != -1) postorder(lc[x]);
        if(rc[x] != -1) postorder(rc[x]);
        cout << (char)(x+65);
    }
};

int main(){
    int n;
    char x, y, z;

    cin >> n;
    Tree T(n);
    for(int i=0; i<n; i++){
        cin >> x >> y >> z;
        T.add_sub_tree(x-65,y-65,z-65);
    }

    T.preorder(0);
    cout << endl;
    T.inorder(0);
    cout << endl;
    T.postorder(0);

    return 0;
}

File: 1900/1991.py
def Preorder(x):
    print(chr(x+65),end='')
    if Left_Children[x]!=-1: Preorder(Left_Children[x])
    if Right_Childeren[x]!=-1:Preorder(Right_Childeren[x])
def Inorder(x):
    if Left_Children[x]!=-1: Inorder(Left_Children[x])
    print(chr(x+65),end='')
    if Right_Childeren[x]!=-1: Inorder(Right_Childeren[x])
def Postorder(x):
    if Left_Children[x]!=-1:Postorder(Left_Children[x])
    if Right_Childeren[x]!=-1:Postorder(Right_Childeren[x])
    print(chr(x+65),end='')
n=int(input())
Left_Children=[-1]*n
Right_Childeren=[-1]*n
for i in range(n):
    a,b,c=input().split()
    if b!='.':Left_Children[ord(a)-65]=ord(b)-65
    if c!='.':Right_Childeren[ord(a)-65]=ord(c)-65
Preorder(0)
print('')
Inorder(0)
print('')
Postorder(0)